### PR TITLE
feat(solr-transforms): Add delete-by-id and /update command dispatcher

### PR DESCRIPTION
**Pull request by niravpi**
### Description

Adds a command dispatcher for Solr's `/update` endpoint and single document delete-by-id support. The dispatcher parses the JSON body, identifies the command type, and delegates to the appropriate handler. Also refactors the response transform from `update-doc.ts` into a generic handler that supports create, update, and delete responses.

**Supported commands on `/update`:**

| Solr command | OpenSearch translation | Status |
|---|---|---|
| `{"delete": {"id": "1"}}` | `DELETE /{collection}/_doc/1` | ✅ New |
| `{"add": {"doc": {...}}}` | `PUT /{collection}/_doc/{id}` (delegates to update-doc) | ✅ New |
| `{"delete": {"query": "..."}}` | — | ❌ Fail fast |
| `{"commit": {}}` | — | ❌ Fail fast |
| Mixed commands | — | ❌ Fail fast |
| Array/bulk operations | — | ❌ Fail fast |
| XML content type | — | ❌ Fail fast |

**Architecture:**

```
POST /solr/{collection}/update
        │
        ▼
  update-command.ts (dispatcher)
        │
        ├── "delete" with "id"  →  delete-doc.ts  →  DELETE /_doc/{id}
        ├── "add" with "doc"    →  update-doc.ts   →  PUT /_doc/{id}
        └── unknown/mixed       →  fail fast
```

```
Pipeline order:
  update: [
    updateDoc.request,      // /update/json/docs (direct path, unchanged)
    updateCommand.request,  // /update (command dispatcher)
  ]
  response: [
    updateCommand.response, // generic _doc response → Solr format
  ]
```

**Response transform refactoring:**
- Moved from `update-doc.ts` to `update-command.ts` since it's generic for all `_doc` API operations
- Now handles `created`, `updated`, `deleted`, and `not_found` results (previously only `created`/`updated`)
- `not_found` maps to status 0 (matching Solr's idempotent delete behavior)

### Files Changed (9)

| File | Change |
|---|---|
| `features/update-command.ts` | New — command dispatcher + generic response transform |
| `features/update-command.test.ts` | New — 30 unit tests |
| `features/delete-doc.ts` | New — delete-by-id handler |
| `features/delete-doc.test.ts` | New — 7 unit tests |
| `update-command.testcase.ts` | New — 10 e2e test cases (5 happy + 5 error) |
| `features/update-doc.ts` | Modified — removed response transform (moved to update-command) |
| `features/update-doc.test.ts` | Modified — removed response transform tests (moved to update-command) |
| `registry.ts` | Modified — wired update-command into update endpoint pipeline |
| `docs/LIMITATIONS.md` | Modified — added UPDATE-COMMANDS section |

### Testing

| Layer | Result |
|---|---|
| Unit tests (vitest) | 26 files, 581 tests ✅ |
| E2E integration (Solr 8 + 9) | 10/10 passed ✅ |
| Sandbox (200K dataset, 167 queries) | 88/167 — same baseline, zero regressions ✅ |
| Manual sandbox validation | 14/14 tests passed ✅ |

### Check List
- [x] New functionality includes testing
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose), if applicable.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: TrafficCapture/SolrTransformations/transforms/src/solr-to-opensearch/features/delete-doc.ts [hunk 1]
```diff
@@ -0,0 +1,24 @@
+/**
```
Comment (k-rooot) [inline TrafficCapture/SolrTransformations/transforms/src/solr-to-opensearch/features/delete-doc.ts:1]: Any thoughts on implementing the update functionalities as a single update-router transform that inspects the request body and dispatches to sub-handlers:

update-router.ts (match: /update/ endpoint)
  ├── detects body shape → routes to:
  ├── update-doc.ts        — single doc add (existing, /update/json/docs)
  ├── update-delete.ts     — delete by id / delete by query
  ├── update-bulk.ts       — array body → _bulk API
  ├── update-commit.ts     — commit/optimize commands
  └── update-atomic.ts     — partial updates with modifiers

The router would be a single micro transform registered on the update endpoint that inspects the body keys (delete, commit, optimize, or array/object doc) and calls the appropriate handler. This keeps the pipeline registration clean, one entry point for `/update`.

Let me know what you think.

Comment (niravpi) [reply TrafficCapture/SolrTransformations/transforms/src/solr-to-opensearch/features/delete-doc.ts:1]: Thanks for the suggestion! Refactored to implement exactly this pattern, a single `update-router.ts` that serves as the sole entry point for all `/update/*` paths.

   **Architecture:**

   update-router.ts (single entry point for /update/*)
   ├── /update/json/docs             → update-doc.ts (body IS the document)
   ├── /update {"add":{"doc":{…}}}   → update-doc.ts (unwraps add.doc)
   ├── /update {"delete":{"id":"…"}} → delete-doc.ts
   └── unsupported commands          → fail fast with clear error

   **Registry now has a single entry:**
   ```ts
   update: [
     updateRouter.request,   // single entry point: routes /update/* by path and body shape
   ]

   The router detects the path (/update/json/docs vs /update) and body shape (command type), then delegates to the appropriate handler via direct .apply() call.

File: TrafficCapture/SolrTransformations/docs/LIMITATIONS.md [hunk 2]
```diff
@@ -368,3 +369,35 @@ internal heuristics.
...
+`add`, `delete` (by id or query), `commit`, `optimize`, `rollback`. Commands
+can be mixed in a single request and can use arrays for bulk operations.
+Both JSON and XML content types are supported.
+
+**Current support:**
+
+| Command | Status |
+|---|---|
+| `add` (single doc) | ✅ Supported — `{"add":{"doc":{...}}}` |
+| `add` with `boost` | ❌ Not supported — boost is silently ignored (OpenSearch has no document-level boost) |
```
Comment (k-rooot) [inline TrafficCapture/SolrTransformations/docs/LIMITATIONS.md]: Instead of silently dropping the boost, we should fail the request with a clear reason such as “unsupported in OpenSearch” or “translation unavailable.” This aligns with the fail-fast approach we are adopting for query translation.

File: TrafficCapture/SolrTransformations/docs/LIMITATIONS.md [hunk 3]
```diff
@@ -368,3 +369,35 @@ internal heuristics.
...
+| `add` with `overwrite: false` | ❌ Not supported — always overwrites |
+| `delete` by id | ✅ Supported — `{"delete":{"id":"..."}}` |
```
Comment (k-rooot) [inline TrafficCapture/SolrTransformations/docs/LIMITATIONS.md]: Will these two fail silently or with appropriate messaging?

File: TrafficCapture/SolrTransformations/transforms/src/solr-to-opensearch/features/update-router.ts [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+  return /\/update\/json\/docs/.test(uri);
+}
+
+export const request: MicroTransform<RequestContext> = {
+  name: 'update-router',
+  apply: (ctx) => {
+    const uri = ctx.msg.get('URI') || '';
+
+    // /update/json/docs — body IS the document, delegate directly
+    if (isJsonDocsPath(uri)) {
```
Comment (k-rooot) [inline TrafficCapture/SolrTransformations/transforms/src/solr-to-opensearch/features/update-router.ts:96]: This path skips all validation performed in dispatchCommand.
* /update validates empty body, unsupported formats
* /update/json/docs blindly forwards

This can lead to subtle bugs (e.g., empty body or array payloads slipping through).

File: TrafficCapture/SolrTransformations/transforms/src/solr-to-opensearch/features/update-router.ts [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+
+  const command = identifyCommand(body);
+  if (!command) {
+    throw new Error('[update-router] No recognized command in request body');
+  }
+
+  const commandData = body.get(command);
+
+  // Array/bulk not supported — single command only
+  if (Array.isArray(commandData) || commandData?.get?.('0') !== undefined) {
```
Comment (k-rooot) [inline TrafficCapture/SolrTransformations/transforms/src/solr-to-opensearch/features/update-router.ts]: This check is fragile for two reasons:
* get('0') depends on Map semantics and will not work with plain objects
* It does not correctly detect all array cases (especially nested structures like {"add":[{"doc":...}]})

Any thoughts to normalise first, then rely on Array.isArray() only.

File: TrafficCapture/SolrTransformations/transforms/src/solr-to-opensearch/features/update-router.ts [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+    throw new Error('[update-router] Request body is empty or not JSON — only JSON content type is supported');
+  }
+
+  if (hasMultipleCommands(body)) {
+    throw new Error('[update-router] Mixed commands in a single request are not supported yet');
+  }
+
+  const command = identifyCommand(body);
+  if (!command) {
+    throw new Error('[update-router] No recognized command in request body');
```
Comment (k-rooot) [inline TrafficCapture/SolrTransformations/transforms/src/solr-to-opensearch/features/update-router.ts]: 1. The error gives no visibility into what was actually received. Can we include the keys from the request body in the error message. This will make debugging significantly easier.

2. Across the file, error messages vary in tone and structure:* "Array/bulk operations are not supported yet"* "Add command must contain a \"doc\" field"

Can we standardize the format?
```
[update-router] <operation>: <reason>
```

File: TrafficCapture/SolrTransformations/transforms/src/solr-to-opensearch/features/update-router.ts [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+
+function handleDelete(ctx: RequestContext, data: any): void {
+  if (!data || typeof data.has !== 'function') {
+    throw new Error('[update-router] Invalid delete command format');
+  }
+  if (data.has('query')) {
+    throw new Error('[update-router] Delete-by-query is not supported yet');
+  }
+  ctx.body = data;
+  deleteDocRequest.apply(ctx);
```
Comment (k-rooot) [inline TrafficCapture/SolrTransformations/transforms/src/solr-to-opensearch/features/update-router.ts:159]: This simply forwards the entire delete payload downstream. However, Solr delete supports additional fields:

```
{"delete":{"id":"1","_route_":"x","version":123}}
```

Right now, _route_ is silently ignored and versioning is lost

Even if downstream doesn’t support these yet, silently dropping them is risky. Can we explicitly extract and validate supported fields (id), and fail clearly for unsupported ones

File: TrafficCapture/SolrTransformations/transforms/src/solr-to-opensearch/features/update-router.ts [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+ * Generic for all _doc API operations (create, update, delete).
+ * OpenSearch returns: {"_index":"mycore","_id":"1","result":"created|updated|deleted",...}
+ * Solr returns:       {"responseHeader":{"status":0,"QTime":N}}
+ */
+export const response: MicroTransform<ResponseContext> = {
+  name: 'update-response',
+  match: (ctx) => ctx.responseBody.has('result') && ctx.responseBody.has('_id'),
+  apply: (ctx) => {
+    const result = ctx.responseBody.get('result');
+    const status = (result === 'created' || result === 'updated' || result === 'deleted' || result === 'not_found') ? 0 : 1;
```
Comment (k-rooot) [inline TrafficCapture/SolrTransformations/transforms/src/solr-to-opensearch/features/update-router.ts:201]: Treating not_found as success is questionable. While Solr can treat delete of a missing doc as success, that behaviour is context-dependent and should be explicitly defined not inferred.

Can we consider HTTP status code (if available in context) and document how not_found should be treated.

File: TrafficCapture/SolrTransformations/transforms/src/solr-to-opensearch/features/update-router.ts [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+    for (const key of keys) {
+      ctx.responseBody.delete(key);
+    }
+
+    // QTime is 0 because OpenSearch's _doc response doesn't include processing time.
+    ctx.responseBody.set(
+      'responseHeader',
+      new Map<string, unknown>([
+        ['status', status],
+        ['QTime', 0],
```
Comment (k-rooot) [inline TrafficCapture/SolrTransformations/transforms/src/solr-to-opensearch/features/update-router.ts:213]: This is misleading. Even if OpenSearch doesn’t return timing, hardcoding 0 can confuse monitoring and debugging.

File: TrafficCapture/SolrTransformations/transforms/src/solr-to-opensearch/features/update-router.ts [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+  for (const cmd of KNOWN_COMMANDS) {
+    if (body.has(cmd)) count++;
+    if (count > 1) return true;
+  }
+  return false;
+}
+
+/** Check if the path is the /update/json/docs shortcut. */
+function isJsonDocsPath(uri: string): boolean {
+  return /\/update\/json\/docs/.test(uri);
```
Comment (k-rooot) [inline TrafficCapture/SolrTransformations/transforms/src/solr-to-opensearch/features/update-router.ts]: This regex will match unintended paths like `/foo/update/json/docs123`

File: TrafficCapture/SolrTransformations/transforms/src/solr-to-opensearch/features/update-router.ts [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+ *   /update  mixed commands             → multiple commands in one request
+ *   /update  XML body                   → only JSON supported
+ *
+ * Future handlers are added as new cases in dispatchCommand() —
+ * no pipeline or registry changes needed.
+ *
+ * Response transform is also here — generic for all _doc API responses
+ * (create, update, delete). Converts OpenSearch format to Solr format.
+ */
+import type { MicroTransform } from '../pipeline';
```
Comment (k-rooot) [inline TrafficCapture/SolrTransformations/transforms/src/solr-to-opensearch/features/update-router.ts:29]: 1. There is no explicit check for JSON content type. Given that XML is explicitly unsupported (as per comments), requests with XML payloads will fail in unpredictable ways. Can we validate Content-Type early and fail clearly.

2. Right now the router follows a simple:
```
detect → dispatch
```

This works for P0, but future features (bulk, atomic updates) will require:
```
parse → normalize → dispatch
```

You’re very close to needing this shift, so it’s worth keeping in mind as you extend functionality.

File: TrafficCapture/SolrTransformations/transforms/src/solr-to-opensearch/features/update-router.ts [hunk 12]
```diff
@@ -0,0 +1,161 @@
...
+  }
+
+  const commandData = body.get(command);
+
+  // Array/bulk not supported — single command only
+  if (Array.isArray(commandData) || commandData?.get?.('0') !== undefined) {
+    throw new Error('[update-router] Array/bulk operations are not supported yet — send one command at a time');
+  }
+
+  switch (command) {
```
Comment (k-rooot) [inline TrafficCapture/SolrTransformations/transforms/src/solr-to-opensearch/features/update-router.ts]: This works, but as more commands are added (commit, optimize, bulk), this will grow. Can we replace with a handler registry map for better scalability.

File: TrafficCapture/SolrTransformations/transforms/src/solr-to-opensearch/registry.ts [hunk 13]
```diff
@@ -99,7 +99,7 @@ export const responseRegistry: TransformRegistry<ResponseContext> = {
       responseHeader.response, // synthesize responseHeader
     ],
     update: [
-      updateDoc.response, // OpenSearch _doc response → Solr update response format
+      updateRouter.response,  // generic _doc response → Solr format (create/update/delete)
```
Comment (k-rooot) [inline TrafficCapture/SolrTransformations/transforms/src/solr-to-opensearch/registry.ts:104]: Any reason why we need to register update in two places.

File: TrafficCapture/SolrTransformations/transforms/src/solr-to-opensearch/update-router.testcase.ts [hunk 14]
```diff
@@ -0,0 +1,161 @@
+/**
+ * E2E test cases for /update command dispatcher — delete-by-id and add-via-update.
```
Comment (k-rooot) [inline TrafficCapture/SolrTransformations/transforms/src/solr-to-opensearch/update-router.testcase.ts:2]: 1. The below tests validate response shape but not actual side effects. 

`update-cmd-delete-by-id`, `update-cmd-add-single-doc` -> Right now, these tests only assert the response format, not whether the document was actually deleted or indexed. 

This is a critical gap for an E2E test. eg: A bug in deleteDocRequest could silently fail, and this test would still pass.

2. Missing coverage for `/update/json/docs`

3. Can we add a test for invalid body types (non-JSON) and unknown commands

4. Delete tests don’t cover edge cases beyond happy path -> numeric IDs, long ids, IDs with reserved characters (e.g. /, ?, #)

5. Can we add test for missing doc in add command

File: TrafficCapture/SolrTransformations/transforms/src/solr-to-opensearch/update-router.testcase.ts [hunk 15]
```diff
@@ -0,0 +1,161 @@
...
+  // --- error paths ---
+
+  solrTest('update-cmd-error-delete-by-query', {
+    description: 'Delete-by-query should return 500',
+    documents: [{ id: '1', title: 'test' }],
+    method: 'POST',
+    requestBody: JSON.stringify({ delete: { query: 'title:test' } }),
+    requestPath: '/solr/testcollection/update',
+    expectedStatusCode: 500,
+    expectedErrorContains: 'Request transform failed',
```
Comment (k-rooot) [inline TrafficCapture/SolrTransformations/transforms/src/solr-to-opensearch/update-router.testcase.ts:109]: This pattern of generic messaging validation is repeated across all error tests. is this intentional? Should  we validate which failure occurred?

File: TrafficCapture/SolrTransformations/transforms/src/solr-to-opensearch/features/update-router.test.ts [hunk 16]
```diff
@@ -0,0 +1,247 @@
...
+}
+
+function addBody(doc: Record<string, any>): Map<string, any> {
+  return new Map([['add', new Map([['doc', new Map(Object.entries(doc))]])]]);
+}
+
+describe('update-router request', () => {
+  // --- routing ---
+
+  it('routes /update/json/docs to update-doc handler', () => {
```
Comment (k-rooot) [inline TrafficCapture/SolrTransformations/transforms/src/solr-to-opensearch/features/update-router.test.ts:35]: Should we verify that the body is preserved correctly and that fields are passed through as expected

File: TrafficCapture/SolrTransformations/transforms/src/solr-to-opensearch/features/update-router.test.ts [hunk 17]
```diff
@@ -0,0 +1,247 @@
...
+    const ctx: RequestContext = {
+      msg, endpoint: 'update', collection: 'mycore',
+      params: new URLSearchParams(), body: body as unknown as JavaMap,
+    };
+    request.apply(ctx);
+    expect(ctx.msg.get('URI')).toBe('/mycore/_doc/1');
+    expect(ctx.msg.get('method')).toBe('PUT');
+  });
+
+  it('routes /update with delete command to delete-doc handler', () => {
```
Comment (k-rooot) [inline TrafficCapture/SolrTransformations/transforms/src/solr-to-opensearch/features/update-router.test.ts:51]: Seems like a duplicate of `dispatches delete to DELETE /_doc/{id}`.

File: TrafficCapture/SolrTransformations/transforms/src/solr-to-opensearch/features/update-router.test.ts [hunk 18]
```diff
@@ -0,0 +1,247 @@
...
+  });
+
+  it('throws on unsupported optimize command', () => {
+    const ctx = buildCtx('/solr/mycore/update', new Map([['optimize', new Map()]]));
+    expect(() => request.apply(ctx)).toThrow("'optimize' is not supported yet");
+  });
+
+  it('throws on array of deletes', () => {
+    // Simulate Java ArrayList with numeric keys + length
+    const deleteArray = new Map<string, any>([['0', new Map([['id', '1']])], ['1', new Map([['id', '2']])], ['length', 2]]);
```
Comment (k-rooot) [inline TrafficCapture/SolrTransformations/transforms/src/solr-to-opensearch/features/update-router.test.ts]: This is a clever hack, but tightly coupled to current implementation (get('0'))  and not representative of real JS arrays

File: TrafficCapture/SolrTransformations/transforms/src/solr-to-opensearch/features/update-router.test.ts [hunk 19]
```diff
@@ -0,0 +1,247 @@
+import { describe, it, expect } from 'vitest';
+import { request, response } from './update-router';
+import type { RequestContext, ResponseContext, JavaMap } from '../context';
+
+/** Build a RequestContext with a nested Map body (auto-converts objects to Maps). */
```
Comment (k-rooot) [inline TrafficCapture/SolrTransformations/transforms/src/solr-to-opensearch/features/update-router.test.ts]: Can we add test coverage for 
1. unknown command payload
2. malformed add.doc
3. /update with no command but valid JSON
4. response transform doesn’t test partial/malformed responses
5. unexpected result values
6. response cleanup completeness

File: TrafficCapture/SolrTransformations/transforms/src/solr-to-opensearch/features/delete-doc.ts [hunk 20]
```diff
@@ -0,0 +1,24 @@
...
+ * OpenSearch:  DELETE /{collection}/_doc/1
+ */
+import type { MicroTransform } from '../pipeline';
+import type { RequestContext } from '../context';
+
+export const request: MicroTransform<RequestContext> = {
+  name: 'delete-doc',
+  apply: (ctx) => {
+    const id = ctx.body.get('id');
+    if (id == null || id === '') {
```
Comment (k-rooot) [inline TrafficCapture/SolrTransformations/transforms/src/solr-to-opensearch/features/delete-doc.ts]: This allows `id = '   '` which becomes `/_doc/%20%20%20`. Should we trim?

File: TrafficCapture/SolrTransformations/transforms/src/solr-to-opensearch/features/delete-doc.ts [hunk 21]
```diff
@@ -0,0 +1,24 @@
...
+      throw new Error('[delete-doc] Delete command must have a non-empty "id" field');
+    }
+
+    ctx.msg.set('URI', `/${ctx.collection}/_doc/${encodeURIComponent(String(id))}`);
```
Comment (k-rooot) [inline TrafficCapture/SolrTransformations/transforms/src/solr-to-opensearch/features/delete-doc.ts]: If collection is undefined `/undefined/_doc/1`. Should we add a validation here?

File: TrafficCapture/SolrTransformations/transforms/src/solr-to-opensearch/features/delete-doc.ts [hunk 22]
```diff
@@ -0,0 +1,24 @@
...
+    ctx.msg.set('method', 'DELETE');
```
Comment (k-rooot) [inline TrafficCapture/SolrTransformations/transforms/src/solr-to-opensearch/features/delete-doc.ts:34]: 1. As per Solr behaviour, `{"delete":{"id":"1","_route_":"x","version":123}}`

_route_ may be required for correct shard routing and versioning may matter for concurrency. Even if not supported, silently ignoring is risky.

2. is it intentional to overwrite method unconditionally?

**Review by k-rooot — CHANGES_REQUESTED** [review]

File: TrafficCapture/SolrTransformations/docs/LIMITATIONS.md (hunk 2 above)
Comment (niravpi) [reply TrafficCapture/SolrTransformations/docs/LIMITATIONS.md]: Fixed `handleAdd` now throws `"[update-router] add: document-level boost is not supported — OpenSearch has no equivalent"`. Updated LIMITATIONS.md. Added unit test + e2e test.

File: TrafficCapture/SolrTransformations/docs/LIMITATIONS.md (hunk 3 above)
Comment (niravpi) [reply TrafficCapture/SolrTransformations/docs/LIMITATIONS.md]: Fixed  throws `"[update-router] add: overwrite=false is not supported — OpenSearch always overwrites"` when `overwrite` is explicitly `false`. `overwrite=true` passes through. Updated LIMITATIONS.md. Added unit test + e2e test.

File: TrafficCapture/SolrTransformations/transforms/src/solr-to-opensearch/features/update-router.ts (hunk 4 above)
Comment (niravpi) [reply TrafficCapture/SolrTransformations/transforms/src/solr-to-opensearch/features/update-router.ts:96]: Fixed, added empty body validation in the router before delegating to `/update/json/docs`. The router now throws `"[update-router] json-docs: request body is empty"` for empty/missing bodies, acting as a gatekeeper. `update-doc.ts` still has its own validation as a defense-in-depth layer.

File: TrafficCapture/SolrTransformations/transforms/src/solr-to-opensearch/features/update-router.ts (hunk 5 above)
Comment (niravpi) [reply TrafficCapture/SolrTransformations/transforms/src/solr-to-opensearch/features/update-router.ts]: Replaced `commandData?.get?.('0')` with `isArrayLike()` helper using `Symbol.iterator` check. Excludes strings and Maps. Works for both JS arrays and GraalVM Java ArrayLists.

File: TrafficCapture/SolrTransformations/transforms/src/solr-to-opensearch/features/update-router.ts (hunk 6 above)
Comment (niravpi) [reply TrafficCapture/SolrTransformations/transforms/src/solr-to-opensearch/features/update-router.ts]: >The error gives no visibility into what was actually received. Can we include the keys from the request body in the error message. This will make debugging significantly easier.

Added `bodyKeys()` helper. Unrecognized command error now includes actual keys: `"[update-router] dispatch: no recognized command in request body (keys: foo, baz)"`.

>Across the file, error messages vary in tone and structure:* "Array/bulk operations are not supported yet"* "Add command must contain a "doc" field"

All error messages now follow [update-router] <operation>: <reason> pattern consistently. Examples: [update-router] delete: delete-by-query is not supported yet, [update-router] dispatch: request body is empty or not JSON, [update-router] add: document-level boost is not supported.

File: TrafficCapture/SolrTransformations/transforms/src/solr-to-opensearch/features/update-router.ts (hunk 7 above)
Comment (niravpi) [reply TrafficCapture/SolrTransformations/transforms/src/solr-to-opensearch/features/update-router.ts:159]: Ack, handleDelete now explicitly validates that only supported fields are present. Added a SUPPORTED_DELETE_FIELDS Set containing ['id']. The handler iterates all keys in the delete payload and fails fast on any unsupported field:  [update-router] delete: unsupported field '_route_' only 'id' is supported
This ensures Solr-specific fields like _route_, version, and commitWithin are never silently dropped. As we add support for more fields, we simply add them to the Set.

File: TrafficCapture/SolrTransformations/transforms/src/solr-to-opensearch/features/update-router.ts (hunk 8 above)
Comment (niravpi) [reply TrafficCapture/SolrTransformations/transforms/src/solr-to-opensearch/features/update-router.ts:201]: In Solr, deleting a non-existent doc returns `{"responseHeader":{"status":0}}`. This is Solr's documented idempotent delete behavior.

Regarding HTTP status code: OpenSearch returns HTTP 404 with `result: "not_found"` for non-existent docs, vs HTTP 200 with `result: "deleted"` for existing docs. The status code is available in the response map via `ctx.response.get('code')`, though we currently use the `result` field for simplicity since it provides the same information.
Added detailed code comment documenting the explicit result mapping:
  > created/updated/deleted → status 0 (success)
  > not_found → status 0 (Solr treats delete of missing doc as success — verified against Solr 8/9)
  > anything else → status 1 (error)

This is an explicit, documented decision matching Solr's actual behavior, not an inferred default.

File: TrafficCapture/SolrTransformations/transforms/src/solr-to-opensearch/features/update-router.ts (hunk 9 above)
Comment (niravpi) [reply TrafficCapture/SolrTransformations/transforms/src/solr-to-opensearch/features/update-router.ts:213]: Agree that hardcoding 0 is not ideal. However, OpenSearch's _doc API response doesn't include any processing time field — unlike the _search API which returns took. We have three options:
   1. `QTime: 0` (current) — technically inaccurate but Solr clients expect a number
   2. `QTime: -1` — signals "not available" but may break clients that assume non-negative
   3. Omit `QTime` — but Solr always includes it, and some clients may fail on missing field
We went with 0 to match the pattern used by the existing response-header.ts for select queries (which also sets QTime: 0). Added a code comment making this explicit.

Open to changing to -1 if the team prefers a clearer "not available" signal, but that would need to be consistent across all response transforms (select also uses 0 today). We can take that up in follow up CR if needed

File: TrafficCapture/SolrTransformations/transforms/src/solr-to-opensearch/features/update-router.ts (hunk 10 above)
Comment (niravpi) [reply TrafficCapture/SolrTransformations/transforms/src/solr-to-opensearch/features/update-router.ts]: Fixed, tightened regex from /\/update\/json\/docs/ to /\/update\/json\/docs(\?|$)/. Now only matches paths ending with /docs or /docs?... (query string). Paths like /update/json/docs123 no longer match. Added
unit test verifying this.

File: TrafficCapture/SolrTransformations/transforms/src/solr-to-opensearch/features/update-router.ts (hunk 11 above)
Comment (niravpi) [reply TrafficCapture/SolrTransformations/transforms/src/solr-to-opensearch/features/update-router.ts:29]: Added Content-Type validation at the top of the router, before any routing logic. Non-JSON content types now fail fast: [update-router] dispatch: only JSON content type is supported, got 'text/xml'

 Regarding the ```parse → normalize → dispatch pattern``` agreed this will be needed for bulk and atomic updates. The current detect → dispatch is sufficient for single-command operations, but the handler registry
   map we introduced (COMMAND_HANDLERS) is designed with this evolution in mind. When we add bulk support, the normalization step would sit between command identification and handler dispatch extracting
   individual operations from arrays before routing each to the appropriate handler. No architectural changes needed, just an additional step in the pipeline.

File: TrafficCapture/SolrTransformations/transforms/src/solr-to-opensearch/features/update-router.ts (hunk 12 above)
Comment (niravpi) [reply TrafficCapture/SolrTransformations/transforms/src/solr-to-opensearch/features/update-router.ts]: Done, replaced the switch statement with a COMMAND_HANDLERS registry map

File: TrafficCapture/SolrTransformations/transforms/src/solr-to-opensearch/registry.ts (hunk 13 above)
Comment (niravpi) [reply TrafficCapture/SolrTransformations/transforms/src/solr-to-opensearch/registry.ts:104]: The pipeline has two separate registries requestRegistry for transforming incoming requests and responseRegistry for transforming outgoing responses. They have different type signatures (RequestContext vs ResponseContext) and can't be combined.

   This is the same pattern used by select, which has 8 request transforms and 5 response transforms registered separately. Every endpoint that transforms both directions needs entries in both registries.

File: TrafficCapture/SolrTransformations/transforms/src/solr-to-opensearch/update-router.testcase.ts (hunk 14 above)
Comment (niravpi) [reply TrafficCapture/SolrTransformations/transforms/src/solr-to-opensearch/update-router.testcase.ts:2]: Added side-effect verification tests using requestSequence — these perform the write operation, then query both Solr and OpenSearch via the shim to verify the actual data:

   - update-cmd-add-then-query: adds a doc via /update, then queries to verify it exists and matches Solr
   - update-cmd-delete-then-query: seeds a doc, deletes via /update, then queries to verify it's gone in both backends

Also found and fixed a bug during this delete-doc.ts wasn't translating ```commit=true → ?refresh=true```, so deleted docs weren't immediately unsearchable in OpenSearch. So thanks for pointing this out.

File: TrafficCapture/SolrTransformations/transforms/src/solr-to-opensearch/features/update-router.test.ts (hunk 16 above)
Comment (niravpi) [reply TrafficCapture/SolrTransformations/transforms/src/solr-to-opensearch/features/update-router.test.ts:35]: Intentional, the Java shim wraps all transform errors into a generic response: Primary target 'opensearch' failed: Request transform failed for target 'opensearch'

   If we want more specific error matching in e2e, that would require changing the Java shim's error response format to include the original error message which can be taken as a separate enhancement PR.

File: TrafficCapture/SolrTransformations/transforms/src/solr-to-opensearch/features/update-router.test.ts (hunk 17 above)
Comment (niravpi) [reply TrafficCapture/SolrTransformations/transforms/src/solr-to-opensearch/features/update-router.test.ts]: Ack, removed the duplicate. Replaced it with a body preservation test that verifies document fields (`title`, `price`) are correctly passed through when routing `/update/json/docs` to the update-doc handler.

File: TrafficCapture/SolrTransformations/transforms/src/solr-to-opensearch/features/update-router.test.ts (hunk 18 above)
Comment (niravpi) [reply TrafficCapture/SolrTransformations/transforms/src/solr-to-opensearch/features/update-router.test.ts]: Ack, changed to use a real JS array instead of simulating with Map numeric keys

File: TrafficCapture/SolrTransformations/transforms/src/solr-to-opensearch/features/update-router.test.ts (hunk 19 above)
Comment (niravpi) [reply TrafficCapture/SolrTransformations/transforms/src/solr-to-opensearch/features/update-router.test.ts]: Ack, Added all

File: TrafficCapture/SolrTransformations/transforms/src/solr-to-opensearch/features/delete-doc.ts (hunk 20 above)
Comment (niravpi) [reply TrafficCapture/SolrTransformations/transforms/src/solr-to-opensearch/features/delete-doc.ts]: Ack, added String(rawId).trim() in delete-doc.ts. Whitespace-only ids now fail fast with "non-empty id" error. Added unit test for '   ' id.

File: TrafficCapture/SolrTransformations/transforms/src/solr-to-opensearch/features/delete-doc.ts (hunk 21 above)
Comment (niravpi) [reply TrafficCapture/SolrTransformations/transforms/src/solr-to-opensearch/features/delete-doc.ts]: Ack, added if (!ctx.collection) throw guard at the top of delete-doc.ts

File: TrafficCapture/SolrTransformations/transforms/src/solr-to-opensearch/features/delete-doc.ts (hunk 22 above)
Comment (niravpi) [reply TrafficCapture/SolrTransformations/transforms/src/solr-to-opensearch/features/delete-doc.ts:34]: Two parts:

   1. `_route_`/`version` :  Fixed. handleDelete now validates only supported fields using SUPPORTED_DELETE_FIELDS Set. Any unsupported field fails fast: "[update-router] delete: unsupported field '_route_' — only 'id' is supported".
   2. Method overwrite : Yes, intentional. Solr's /update only accepts POST. We convert to the OpenSearch equivalent: DELETE for delete operations, PUT for add operations. Added POST method validation in the router non-POST requests now fail fast: "[update-router] dispatch: only POST method is accepted for /update, got 'GET'".

**Review by k-rooot — APPROVED** [review]
